extensions/amp-auto-ads/amp-auto-ads-spec.md

```diff
@@ -0,0 +1,268 @@
+limitations under the License.
+-->
+
+# <a name="amp-auto-ads-spec"></a> `amp-auto-ads` spec
+
+## Description
+`amp-auto-ads` uses a remotely served configuration to dynamically inject ads
+into the page.
+
+## Ad Placement Strategy
+Given a sufficient number of valid placements (supplied in the configuration),
+`amp-auto-ads` will try to insert additional ads within the following
+constraints:
+<ul>
+  <li>No more than 3 ads on the page (including any existing ads)</li>
+  <li>No injected ad within 500px (measured vertically) of another ad</li>
+  <li>
+    Ad can be injected without causing unacceptable re-flow (as determined by
+    attemptChangeSize)
+  </li>
+</ul>
+
+## Configuration
+The configuration defines where on the page amp-auto-ads can place ads. It is
+fetched from a 3rd party ad network at the URL defined in
+`ad-network-config.js`. It should be a serialized JSON object matching the
+`ConfigObj` definition below.
+
+### Object Definitions
+
+#### ConfigObj
+<table>
+  <tr>
+    <th>Field Name</th>
+    <th>Required</th>
+    <th>Type</th>
+    <th>Default</th>
+    <th>Description</th>
+  </tr>
+  <tr>
+    <td>placements</td>
+    <td>Yes</td>
+    <td>Array&lt;!PlacementObj&gt;</td>
+    <td></td>
+    <td>The potential places where ads can be inserted on the page.</td>
+  </tr>
+</table>
+
+#### PlacementObj
+<table>
+  <tr>
+    <th>Field Name</th>
+    <th>Required</th>
+    <th>Type</th>
+    <th>Default</th>
+    <th>Description</th>
+  </tr>
+  <tr>
+    <td>anchor</td>
+    <td>Yes</td>
+    <td>AnchorObj</td>
+    <td></td>
+    <td>
+      Information used to look up the element(s) on the page that the
+      placement position is anchored to.
+    </td>
+  </tr>
+  <tr>
+    <td>pos</td>
+    <td>Yes</td>
+    <td>RelativePositionEnum</td>
+    <td></td>
+    <td>The position of the placement relative to its anchor element.</td>
+  </tr>
+  <tr>
+    <td>type</td>
+    <td>Yes</td>
+    <td>PlacementTypeEnum</td>
+    <td></td>
+    <td>The type of placement.</td>
+  </tr>
+  <tr>
+    <td>style</td>
+    <td>No</td>
+    <td>PlacementStyleObj</td>
+    <td>{}</td>
+    <td>
+      Any styling that should be applied to an ad inserted in this placement
+      position.
+    </td>
+  </tr>
+</table>
+
+#### AnchorObj
+<table>
+  <tr>
+    <th>Field Name</th>
+    <th>Required</th>
+    <th>Type</th>
+    <th>Default</th>
+    <th>Description</th>
+  </tr>
+  <tr>
+    <td>selector</td>
+    <td>Yes</td>
+    <td>string</td>
+    <td></td>
+    <td>
+      A CSS selector to select the element(s) at this level of the anchor
+      definition.
+    </td>
+  </tr>
+  <tr>
+    <td>index</td>
+    <td>No</td>
+    <td>number</td>
+    <td>0 (if all field not set to true)</td>
+    <td>
+      The index of the elements selected by the selector that this level of the
+      anchor definition should be limited to.
+    </td>
+  </tr>
+  <tr>
+    <td>all</td>
```

Comment:
It would be a pain to change for us now, since we have the serving stack in place based on the API that's been submitted in the AMP project so far.

I'm also a little wary if having default behavior based on the non-presence of number fields. There are various data formats that collapse 0 values to the field just not being present (protobuf being one of them...) So it might be one of those things that causes nasty bugs for people trying to interface with this API.